src/docs/arch-timestamp.dox

```diff
@@ -0,0 +1,280 @@
+
+@section arch-timestamp-representation Representation of timestamps and time windows
+
+Internally, timestamps are 64-bit unsigned integers, where 0 is
+reserved and increasing values correspond to later time.
+WiredTiger does not otherwise interpret the values of timestamps and
+they need not be real time or any clock time; sequential counters are
+perfectly satisfactory.
+The symbolic values \c WT_TS_NONE and \c WT_TS_MAX give 0 and the
+maximum possible timestamp, respectively.
+
+Timestamps are exchanged with the application as strings, printed in
+hex without a leading \c 0x, and not as machine integers.
+(This is because they are handled using the configuration parameter
+system, which is string-based.)
+
+Time <i>windows</i> are handled in memory using the struct type
+\c WT_TIME_WINDOW.
+This contains a start time and a stop time.
+The type actually has seven members:
+
+- \c start_ts (the start time)
+- \c durable_start_ts (also the start time)
+- \c start_txn (the transaction ID that committed the start of the value)
+- \c stop_ts (the stop time)
+- \c durable_stop_ts (also the stop time)
+- \c stop_txn (the transaction ID that committed the stop/end of the value)
+- \c prepare (the transaction-prepare state of the value)
+
+The prepare state is not properly speaking part of the time window,
+but it is also a property of the value and needs to be seen at most of
+the same points, and in particular also needs to be stored on disk
+along with the time information, so is kept in the time window
+structure for convenience.
+
+The distinction between \c start_ts and \c durable_start_ts (and
+between \c stop_ts and \c durable_stop_ts) is the same as the
+distinction between a transaction's commit timestamp and its durable
+timestamp.
+See @ref arch-timestamp-prepare.
+
+Start times from at or before the pinned timestamp are all equivalent
+to "no start time" and can be represented as \c WT_TS_NONE.
+Similarly, transaction IDs from past write generations should not be
+retained and are represented as \c WT_TXN_NONE.
+
+A value that is current and has not been overwritten or deleted has no
+stop time.
+This is represented with \c WT_TS_NONE in \c durable_stop_ts but
+\c WT_TS_MAX in \c stop_ts, and \c WT_TS_MAX in \c stop_txn.
+
+The on-disk representation of the time window of the current on-disk
+value is an optional part of a value cell.
+Time windows that are not present are universal (no start time, no
+stop time); the stop times that mean "no stop time" are not stored.
+Other stop times are stored as differences from the start time to save
+space.
+Similarly, durable timestamps are stored as differences from their
+corresponding commit timestamps.
+
+The on-disk representation of the time windows of older values are
+stored differently because those times are used for lookup purposes.
+See @ref arch-hs.
+
+Internal btree pages store aggregated time window information for the
+pages or sub-trees they reference.
```

Comment:
I'm not sure; I think it follows readily from the names of the fields... at least once you're told which ones default to NONE and which to MAX... what's definitely not obvious is why this particular set of fields was chosen (with these particular defaults), but that part I don't know. Anyway I put a little more in. I could make it all very explicit by listing out how each field is aggregated, but to me that feels like overkill.

Reply:
Ah.  I didn't really mean the mechanics of how the aggregate timestamp gets computed but rather what an aggregate time window is or what it means.  In other words, maybe provide a higher-level definition than the list of fields.  "An aggregated time window is the smallest time window that spans the time windows for all of the records in the corresponding page or subtree"? 

I don't think this is a big deal, but I'm thinking a naïve reader might assume a different type of aggregation. Maybe it could mean the intersection of all of the underlying time windows?

Reply:
Ah, completely misunderstood what you wanted. Pushed a change here: https://github.com/sauclovian-wt/wiredtiger/commit/233a58e79aa7ceaac18158e00c351c8deb7cc7b9 and with luck it's in time to get rolled in with the others.

Reply:
Fixed in d1e06e0b8.

Reply:
LGTM.  Thanks @sauclovian-wt & @keithbostic!